### PR TITLE
Make `_cond_test` a method for library-based types

### LINKED ISSUE #17633: `import Regex` breaks ability to write `if myre.match(...)`
### Summary of Problem

As a user who prefers `import` to `use`, compiling and running the following program:

```chapel
import Regex;

var myre=Regex.compile("hello[0-9]");
if myre.match("hello0") then writeln("Yes!");
```

results in an error:

```
testit.chpl:4: error: type 'regexMatch' used in if or while condition
```

This is because, in the current compiler and language definition, support for using an arbitrary type in a conditional is implemented using a (not user-facing) standalone `_cond_test()` routine that is used in conditionals to convert types into booleans.  The nature of this failure is that `_cond_test(:regexMatch)` is implemented in `Regex` and is not brought in by the `import` statement, breaking the conditional statement.

Resolving this issue may involve further design and discussion about how (and whether) users can permit arbitrary types to be used in conditional contexts; potentially converting `_cond_test()` to a user-facing feature.  It feels related to recent discussions on implicit conversions and operator methods.

### PR DESCRIPTION
**Pull request by bradcray**
This PR addresses an issue in which `import`ing a module with a type that can be used in conditionals could cause those conditionals to break.  The reason for the issue is that we implement such checks as standalone functions (named `_cond_test`) and such standalone functions aren't brought into scope for `import` clauses unless they're named.

The fix I took here (which came out of a conversation with @mppf and @lydia-duncan) was to make these routines methods on the types in question (named `chpl_cond_test_method` so that any instance of the object would automatically have access to the method.  To minimize the number of code changes required, I took the approach of checking for the presence of the method in the error overload of `_cond_test` and calling it if it existed (issuing the usual error if it does not).  This avoided changes to the compiler and to the existing `_cond_test` implementations for automatic types (bools, ints, strings, etc.)

Going forward, we will want to consider whether user-defined types should be able to define themselves as being legal to use in a conditional which I've spun off into https://github.com/chapel-lang/chapel/issues/17645.

Resolves #17633

### CODE REVIEW

**Review by vasslitvinov — APPROVED** [review]
Looks good.

File: test/modules/bradc/printModStuff/foo.good [hunk 1]
```diff
@@ -25,14 +25,14 @@ end of module search dirs
   $CHPL_HOME/modules/standard/ChapelEnv.chpl
   $CHPL_HOME/modules/standard/CPtr.chpl
   $CHPL_HOME/modules/standard/HaltWrappers.chpl
+  $CHPL_HOME/modules/standard/Reflection.chpl
```
Comment (mstrout) [inline test/modules/bradc/printModStuff/foo.good:28]: Why did this change happen?

Comment (bradcray) [reply test/modules/bradc/printModStuff/foo.good:28]: This test prints out the files that are parsed when compiling a Chapel program, where the order in which they are parsed depends on the order in which they are `use`d/`import`ed.  I added a new `use Reflection` statement to `ChapelBase.chpl`, which is one of the very earliest modules to be parsed, so it moved the parsing of Reflection earlier in the compilation process than it had been before.

Comment (mstrout) [reply test/modules/bradc/printModStuff/foo.good:28]: Makes sense.  Thanks!

**Pull request merged**